components/device-mgt/org.wso2.carbon.device.mgt.core/src/test/java/org/wso2/carbon/device/mgt/core/authorization/DeviceAccessAuthorizationServiceTest.java

```diff
@@ -0,0 +1,291 @@
+            Assert.assertTrue(deviceAccessAuthorizationService.isUserAuthorized(deviceId, ADMIN_USER),
+                    "Device access authorization for admin user failed");
+        }
+    }
+
+    @Test
+    public void userAuthDevIdUserNameResult() throws Exception {
+        DeviceAuthorizationResult deviceAuthorizationResult = deviceAccessAuthorizationService.
+                isUserAuthorized(deviceIds, ADMIN_USER);
+        Assert.assertEquals(deviceAuthorizationResult.getAuthorizedDevices().size(), 5,
+                "Expected 5 authorized devices for admin user");
+        Assert.assertEquals(deviceAuthorizationResult.getUnauthorizedDevices().size(), 0,
+                "Expected 0 un-authorized devices for admin user");
+    }
+
+    @Test
+    public void userAuthDevId() throws Exception {
+        for (DeviceIdentifier deviceId : deviceIds) {
+            Assert.assertTrue(deviceAccessAuthorizationService.isUserAuthorized(deviceId),
+                    "Authorize user from device identifier failed");
+        }
+    }
+
+    @Test
+    public void userAuthDevIdResult() throws Exception {
+        DeviceAuthorizationResult deviceAuthorizationResult = deviceAccessAuthorizationService.isUserAuthorized(deviceIds);
+        Assert.assertEquals(deviceAuthorizationResult.getAuthorizedDevices().size(), 5,
+                "Expected 5 authorized devices for admin user");
+        Assert.assertEquals(deviceAuthorizationResult.getUnauthorizedDevices().size(), 0,
+                "Expected 0 un-authorized devices for admin user");
+    }
+
+    @Test
+    public void userAuthDevIdPermission() throws DeviceAccessAuthorizationException, UserStoreException, PermissionManagementException {
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(NON_ADMIN_ALLOWED_USER);
+        Assert.assertTrue(deviceAccessAuthorizationService.isUserAuthorized(deviceIds.get(0), new String[]{NON_ADMIN_PERMISSION}),
+                "Non admin user with permissions attempt to access failed");
+        PrivilegedCarbonContext.endTenantFlow();
+    }
+
+    @Test
+    public void userAuthFalseDevIdPermission() throws DeviceAccessAuthorizationException, UserStoreException,
+            PermissionManagementException {
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(MultitenantConstants.SUPER_TENANT_ID, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(NON_ADMIN_ALLOWED_USER);
+        Assert.assertFalse(deviceAccessAuthorizationService.isUserAuthorized(deviceIds.get(3), new String[]{NON_ADMIN_PERMISSION}),
+                "Non admin user accessing not allowed device authorized");
+        PrivilegedCarbonContext.endTenantFlow();
+    }
```

Comment:
Why we are starting and ending tenant flow in all the test cases? Shall we move the start tenant flow to @BeforeTest method and shall we end tenant flow in the tear down ?